### PR TITLE
IA-1250 flush log file

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-1250

Depends on https://github.com/DataBiosphere/leonardo/pull/1085 (has been merged)

1. Some changes are due to google2 upgrade
2. Some changes are for logging exception with traceId
3. Main change is sending welder.log to gcs
4. fix https://broadworkbench.atlassian.net/browse/IA-1454

Screen Shot 2019-10-21 at 10 15 26 PM

### CODE REVIEW

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/Main.scala [hunk 1]
```diff
@@ -21,19 +21,20 @@ object Main extends IOApp {
...
       metadataCache <- cachedResource[RelativePath, AdaptedGcsMetadataCache](
         appConfig.pathToGcsMetadataJson,
-        blockingEc,
+        blocker,
         metadata => List(metadata.localPath -> metadata)
       )
-      streams <- initStreams(appConfig, blockingEc, storageLinksCache, metadataCache)
-      _ <- Stream.emits(streams).covary[IO].parJoin(streams.length)
+      streams <- initStreams(appConfig, blocker, storageLinksCache, metadataCache)
+      _ <- Stream.emits(streams).covary[IO].parJoin(6)
```
Comment (akarukappadath) [inline server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/Main.scala]: this seems like a magic number

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  val service: HttpRoutes[IO] = HttpRoutes.of[IO] {
+    case POST -> Root / "flush" =>
+      flush >> NoContent()
+  }
+
+  val flush: IO[Unit] = {
+    val flushStorageLinks = flushCache(config.storageLinksPath, blocker, storageLinksCache)
+    val flushMetadataCache = flushCache(config.metadataCachePath, blocker, metadataCache)
+
+    val flushLogFile = for {
```
Comment (akarukappadath) [inline server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala]: would like to eventually see an autotest in Leo for this!

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      implicit0(ev: ApplicativeAsk[IO, TraceId]) <- IO(ApplicativeAsk.const[IO, TraceId](TraceId(UUID.randomUUID().toString)))
+      blobName = GcsBlobName("welder/welder.log")
+      _ <- googleStorageAlg.fileToGcs(config.logFilePath, GsPath(config.stagingBucketName, blobName))
+    } yield ()
+
+    Logger[IO].info("Shutting down welder") >> Stream(
+      flushMetadataCache,
+      flushStorageLinks,
+      Stream.eval(flushLogFile))
+      .parJoin(3)
```
Comment (akarukappadath) [inline server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala]: another magic number

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/Main.scala (hunk 1 above)
Comment (Qi77Qi) [reply server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/Main.scala]: ah, this is not intended..it should be `streams.length`

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala (hunk 3 above)
Comment (Qi77Qi) [reply server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ShutdownService.scala]: this is not magical number, it should be the length of the `Stream`, which is 3 here

File: server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/StorageLinksService.scala [hunk 4]
```diff
@@ -77,23 +76,29 @@ class StorageLinksService(
...
-    ((Stream.emits(fileBody) through io.file.writeAll[IO](destinationPath, linerBacker.blockingContext, List(StandardOpenOption.CREATE_NEW))).compile.drain)
-      .recoverWith { case _ => logger.info(s"${config.workspaceBucketNameFileName} already exists") } // If file already exists, ignore the failure
+    val safeModeDestinationPath = config.workingDirectory
+      .resolve(safeModeDirectory.path.asPath)
+      .resolve(config.workspaceBucketNameFileName)
+
+    val writeToFile: java.nio.file.Path => IO[Unit] = destinationPath => ((Stream.emits(fileBody) through io.file.writeAll[IO](destinationPath, blocker, List(StandardOpenOption.CREATE_NEW))).compile.drain)
+      .recoverWith { case _ => logger.info(s"${config.workspaceBucketNameFileName} already exists in ${destinationPath}") } // If file already exists, ignore the failure
+
+    (writeToFile(editModeDestinationPath), writeToFile(safeModeDestinationPath)).parSequence_
```
Comment (Qi77Qi) [inline server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/StorageLinksService.scala]: this fixes https://broadworkbench.atlassian.net/browse/IA-1454

**Pull request merged**